ckanext/canada/triggers.py

```diff
@@ -110,12 +125,39 @@ def update_triggers():
             BEGIN
                 IF value IS NOT NULL AND value <> '' AND NOT (value = ANY (choices)) THEN
                     -- \t is used when converting errors to string
-                    RETURN ARRAY[[field_name, 'Invalid choice: "'
+                    RETURN ARRAY[[field_name, 'Invalid choice: {}\uF8FF"'
                         || replace(value, E'\t', ' ') || '"']];
                 END IF;
                 RETURN NULL;
             END;
         ''')
+    # return record with .clean (trimmed value) and .error
+    # (NULL or ARRAY[[value_trimmed]])
+    # .clean must be unnest(.clean) from the return to get the text value
+    lc.action.datastore_function_create(
+        name=u'max_char_error',
+        or_replace=True,
+        arguments=[
+            {u'argname': u'value', u'argtype': u'text'},
+            {u'argname': u'max_chars', u'argtype': u'numeric'},
+            {u'argname': u'field_name', u'argtype': u'text'},
+            {u'argname': u'clean', u'argtype': u'_text', u'argmode': u'out'},
+            {u'argname': u'error', u'argtype': u'_text', u'argmode': u'out'}],
+        rettype=u'record',
+        definition=u'''
+            DECLARE
+                value_trimmed text := trim(both E'\t\n\x0b\x0c\r ' from value);
+            BEGIN
+                IF value IS NOT NULL AND value <> '' AND LENGTH(value_trimmed) > max_chars THEN
+                    error := ARRAY[[field_name, 'This field has a maximum length of {} characters.\uF8FF' || max_chars]];
+                END IF;
+                IF value IS NULL OR value = '' THEN
```

Comment:
empty strings get converted to NULL but all-whitespace strings get trimmed to an empty string? that's not idempotent

Reply:
what should happen with empty strings? just allow them?

Reply:
I think it's better to always store strings in string values (i.e. empty string instead of a mix of empty strings and NULL values) because there's no useful distinction.